pkg/gadgets/bpftrace/tracer.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2023 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !withoutebpf
+
+package tracer
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"os/exec"
+
+	gadgetcontext "github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-context"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/bpftrace/types"
+)
+
+type Config struct {
+	Program string
+}
+
+type Tracer struct {
+	eventCallback func(ev *types.Event)
+	config        Config
+}
+
+func (t *Tracer) Run(gadgetCtx gadgets.GadgetContext) error {
+	params := gadgetCtx.GadgetParams()
+	t.config.Program = params.Get(ParamProgram).AsString()
+	log := gadgetCtx.Logger()
+
+	cmd := exec.Command("bpftrace", "-e", t.config.Program)
+
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return fmt.Errorf("getting stdout pipe: %w", err)
+	}
+
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return fmt.Errorf("getting stdout pipe: %w", err)
+	}
+
+	if err := cmd.Start(); err != nil {
+		return fmt.Errorf("running bpftrace: %w", err)
+	}
+
+	// TODO: how to enrich with the operator!
+	node := os.Getenv("NODE_NAME")
```

Comment:
@flyth while implementing this gadget I found this situation, how to enrich only the node? (this gadget doesn't support filtering by k8s related things, yet)

I tried to implement this in `KubeManager` but I feel it doesn't fit very well: kubemanager provides parameters for filtering, logic for attaching containers, etc. using it to only enrich the node name seems too much. Another idea was to implement an operator that only enriches the node, but I'm not sure either, seems like it'll add overhead as it'll called for all events that are also handled by `KubeManager`. Do you have any opinion on this?

Reply:
It can be done in the `grpc-runtime` directly, as an additional enricher. `runGadget` has the `node` inside GadgetPod, so you could check for the `SetNode()` interface on the prototype and hand over the enricher.

Something like that (maybe a new interface for only the NodeSetter might be needed, then):

```golang
	if parser != nil {
		var enrichers []func(any) error
		ev := gadgetCtx.GadgetDesc().EventPrototype()
		if _, ok := ev.(operators.ContainerInfoSetters); ok {
			enrichers = append(enrichers, func(ev any) error {
				ev.(operators.ContainerInfoSetters).SetNode(pod.node)
				return nil
			})
		}
		
		jsonHandler = parser.JSONHandlerFunc(enrichers...)
		jsonArrayHandler = parser.JSONHandlerFuncArray(pod.node, enrichers...)
	}
```

Reply:
IMHO it should only be overwritten if not set already - if we implement fanning out requests, the info might be already there from the routing node . I'd prefer having it always set by the grpc-runtime, as it can potentially save _a bit_ of bandwidth and CPU time.


```mermaid
flowchart TD
  subgraph Flow
    direction TB
    client[Client]
    node1[Node 1]
    node2[Node 2]
    client <-- "via grpc-runtime\nsets to node 'node1' if not set" --- node1
    node1 <-- "via grpc-runtime\nsets to node 'node2' if not set" --- node2
  end
```

Reply:
I see. Given that this discussion is not too related to this PR, do you agree if we leave it as it's and then we handle in a different one?

Reply:
Absolutely!